Clustering/optimization

cryogenian opened the pull request:
Basically there are two things: 
+ Timestamps are cached, so, advertisement doesn't requre traversing through whole mapdb
+ Updates are cached and broadcasted separately, this in theory should make advertsements handling less talkative and make cluster more consistent. 

I didn't come up with testing update stream from `TimestampedStore` since it's finished when resource is freed, but it's part of the store under resource 🤔

cryogenian commented:
`143.3.1-5cbf672`

wemrysi commented on impl/src/main/scala/quasar/impl/storage/TimestampedStore.scala:
```diff
@@ -37,12 +44,44 @@ final case class TimestampedStore[F[_]: FlatMap: Timer, K, V](
+      val storeStream = Stream.emit[F, TimestampedStore[F, K, V]](store)
+      storeStream.concurrently(store.updates).compile.resource.lastOrError
+    }
+    Resource.liftF(res).flatten.mapK[F](new ~>[F, F] {
```
:golf: `Resource.suspend(res)`. Also, there is nicer syntax for constructing `~>`:
```scala
λ[F ~> F](ContextShift[F].evalOn(pool.unwrap))
```

wemrysi commented on impl/src/main/scala/quasar/impl/storage/TimestampedStore.scala:
```diff
@@ -37,12 +44,44 @@ final case class TimestampedStore[F[_]: FlatMap: Timer, K, V](
+      store <- Concurrent[F].delay(new TimestampedStore(underlying, r, q))
+    } yield {
+      val storeStream = Stream.emit[F, TimestampedStore[F, K, V]](store)
+      storeStream.concurrently(store.updates).compile.resource.lastOrError
```
Why is `updates` stream drained here?

cryogenian replied:
To close update stream when store resource is relessed

wemrysi replied:
Ok, this shouldn't be necessary as there is nothing to close until something pulls on the stream. What this is effectively doing is dropping all of the updates on the floor.

Only components that actually consume this stream need be concerned with interruption.

wemrysi replied:
To interrupt this when the store is finalized, you could create a `boundedNoneTerminated` queue and add a finalizer here that enqueues `None`. This would interrupt the stream for any consumers without draining all the values.

wemrysi commented on impl/src/main/scala/quasar/impl/storage/AntiEntropyStore.scala:
```diff
@@ -74,44 +73,35 @@ final class AntiEntropyStore[F[_]: ConcurrentEffect: ContextShift: Timer, K: Cod
+    // Note, that we don't handle keys aren't presented in advertisement
+    // They are handled when this node sends its advertisement instead.
+    // for every key in advertisement having the most recent modification timestamp map
+    def result(init: Accum, timestamps: Map[K, Long])  = ad.toList.foldM[F, Accum](init){ (acc, v) => (acc, v) match {
```
:golf: You can omit the `(acc, v) => (acc, v) match {` and just match on the arguments as if they were tupled.

wemrysi commented on impl/src/test/scala/quasar/impl/storage/TimestampedStoreSpec.scala:
```diff
@@ -57,15 +61,16 @@ final class TimestampedStoreSpec extends IndexedStoreSpec[IO, String, String] {
         bar <- Timestamped.tagged[IO, String]("bar")
         _ <- us.insert("foo", bar)
-        ts = TimestampedStore(us)
+        ts <- TimestampedStore(us, pool).use(IO(_))
```
It isn't ever safe to capture the value produced by `use`, should lift the computation into `Resource` instead.

djspiewak replied:
Also as a side note, if we *were* capturing the value produced by `use`, we would do it with `IO.pure` rather than `IO.apply`

wemrysi commented on impl/src/main/scala/quasar/impl/storage/TimestampedStore.scala:
```diff
@@ -37,12 +44,44 @@ final case class TimestampedStore[F[_]: FlatMap: Timer, K, V](
+      r <- Ref.of[F, Map[K, Long]](entries.map {
+        case (k, v) => (k, timestamp(v))
+      })
+      store <- Concurrent[F].delay(new TimestampedStore(underlying, r, q))
```
No need to `delay` here.

cryogenian commented:
@wemrysi 

+ Using lambda for functionk
+ `Resource.suspend` (also in anti entropy store) 
+ Removed redunant `match`
+ Removed redundant `delay`
+ Also `liftF`s in tests and use only in the last expression and `IO.pure` instead of `IO`

djspiewak commented on impl/src/main/scala/quasar/impl/storage/AntiEntropyStore.scala:
```diff
@@ -82,8 +82,8 @@ final class AntiEntropyStore[F[_]: ConcurrentEffect: ContextShift: Timer, K: Cod
-    def result(init: Accum, timestamps: Map[K, Long])  = ad.toList.foldM[F, Accum](init){ (acc, v) => (acc, v) match {
-      case ((requesting, returning), (k, incoming)) => for {
+    def result(init: Accum, timestamps: Map[K, Long])  = ad.toList.foldM[F, Accum](init){
+      case (acc@(requesting, returning), (k, incoming)) => for {
```
Style nit: `acc @ (` is preferred in general, since putting spaces around the `@` avoids the occasional ambiguity in the parser related to annotations.

cryogenian commented:
@wemrysi 

+ Using `NoneTerminatedQueue` and finalizer inside of `Resource.make`
+ Wrapped everything in `evalOn`
+ Spaces around `@`

cryogenian commented:
@wemrysi 
+ `evalOn` was removed
